### ISSUE TITLE
Virtual swapchain causes screenshots mismatch for recaptured files.

### ISSUE THREAD

**Issue by locke-lunarg** [LunarG/gfxreconstruct#1122]
`WriteScreenshots` has a `DeviceWaitIdle`. `DeviceWaitIdle` could fix sync issues to generate a correct screenshot, even if it has any tasks that are uncompleted. But in some cases, `virtual swapchain` could cause `DeviceWaitIdle` helpless.

If virtual swapchain is enabled during re-capturing, the virtual swapchain code would be written to the file. And then during replay this recaptured file, it will run the captured virtual swapchain code, like
```
  "index": 335713,
  "function": {
    "name": "vkCmdBlitImage",
    …
      "srcImage": 73,
      "dstImage": 70,
    …
       }
```
`srcImage: 73` is the captured virtual swapchain image. `dstImage: 70` is a new virtual swapchain image. It copies a old virtual swapchain image to a new virtual swapchain image. The old virtual swapchain image could be an uncompleted image if it has sync issues, and then, output a screenshot (image: 70). Even though, `WriteScreenshots` has a `DeviceWaitIdle`, it's behind `vkCmdBlitImage`. It can't help `image: 70` that has already been copied from an uncompleted image. If virtual swapchain is disable during recapturing processing, it wouldn't have this issue. 

However, if the recapturing process is with screenshots, the `WriteScreenshots` code would be also written to the file. And the captured `WriteScreenshots` code is above the captured `virtual swapchain` code, so it would run `DeviceWaitIdle` before `vkCmdBlitImage`, so the new virtual swapchain image would be copied an completed old virtual swapchain image. The screenshot would be good.

`--sync` could fix this issue. But how could we fix this bug in the code?

**Comment by locke-lunarg** [comment]
This issue should be closed. This issue comes from apps that have sync issues. It's not a gfxr issue. It's reasonable to use `--syn` to fix it. I don't think we could have a better idea than `--syn` for this case.

**Comment by bradgrantham-lunarg** [comment]
I am not 100% convinced we don't have some issue with screenshot.  The app seems to work fine on its own, so I want to know for sure that the app has incorrect usage of synchronization under Vulkan and we're doing our synchronization correctly.  Mark volunteered to take a look.

**Comment by MarkY-LunarG** [comment]
I'm wondering if the problem isn't the replay but the capture.  It still might be an application bug, but look at my results from a recent run.  The left is the test run results, the right is the expected image.  On the left, the top file is the "capture" file and the bottom file is the "replay" file.  Notice the replay and the expected match, but the capture doesn't.  This could indicate some kind of sync issue during capture.  

![image](https://github.com/LunarG/gfxreconstruct/assets/16328578/3d46b60d-c93e-4bab-b0a7-9fd8f3cde3eb)

But I'm going to keep looking into this for now.

**Comment by locke-lunarg** [comment]
https://drive.google.com/drive/folders/1G-J7czqG-755Qcva5p5x92JN4qzh4PGZ?usp=drive_link
I have my report about this issue is in this folder.

**Comment by locke-lunarg** [comment]
`ScreenshotHandler::WriteImage` has `DeviceWaitIdle` to wait all command completed whatever the commands are from gfxr or the game, so if the game has sync issues, `DeviceWaitIdle` could cause the the gfxr's screenshot is different to the real game. If we want an accurate screenshot, we might remove `DeviceWaitIdle` to not intervene the game's process.

**Comment by MarkY-LunarG** [comment]
The stuttering issue I see that leads to this goes away if you disable "virtual swapchain".  So this definitely seems like a bug in our Virtual Swapchain implementation.

**Comment by MarkY-LunarG** [comment]
After some investigation, the problem appears that when we use `Virtual Swapchain`, we perform the Blit from the virtual image to the hardware image in a specific queue which we select at `vkCreateSwapchain` time.  The incoming `vkQueuePresent`s from Wolfenstein come in different queues over time, some of which may be different  than what we set up during `vkCreateSwapchain`.

Wolfenstein performs synchronization within it's queues using the last `vkQueueSubmit` before the `vkQueuePresentKHR`.  That `vkQueueSubmit` has all the semaphore and fence waiting logic.  The `vkQueuePresentKHR` follows behind with no semaphore or fence logic so has no sync point itself (other than it's location in the queue after the synchronized submit), but that is still valid.

**Why is this a problem for Virtual Swapchain?**

Since `Virtual Swapchain` uses a specific queue, the commands to Blit the image could get inserted into this other queue.  This wouldn't be a problem if we could grab a semaphore from the `vkQueuePresentKHR`, and wait to do the Blit work until that semaphore has triggered.  However, in this case, we don't have a semaphore so we could start the Blit immediately in the other queue while the work has not been complete in the presenting queue.  This means we have a race condition under Vulkan work structured this way.

**What can we do?**

After some discussion, @locke-lunarg  and I have come up with a solution.  At `vkSwapchainCreate` time, we will create a command buffer for every queue family associated with the Vulkan device.  Then, at queue present time, we will use the command buffer we created that can be associated with the queue used for the `vkQueuePresentKHR`.  This in effect inserts the Blit commands after the synchronized `vkQueueSubmit`, but before the `vkQueuePresentKHR`.  Thus removing the race condition.

Preliminary work has been done, and it appears to fix the issue, but cleanup and testing is still ongoing. 

Thanks to @locke-lunarg for helping with this investigation.
